mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerDBRecord.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.mssql;
+
+import io.cdap.cdap.api.data.format.StructuredRecord;
+import io.cdap.cdap.api.data.schema.Schema;
+import io.cdap.plugin.db.DBRecord;
+import io.cdap.plugin.db.SchemaReader;
+
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Types;
+
+import static io.cdap.plugin.mssql.SqlServerSchemaReader.GEOGRAPHY_TYPE;
+import static io.cdap.plugin.mssql.SqlServerSchemaReader.GEOMETRY_TYPE;
+
+/**
+ * Abstract SQL Server implementation of Source/Sink {@link org.apache.hadoop.mapreduce.lib.db.DBWritable} and
+ * {@link org.apache.hadoop.io.Writable}.
+ * Groups {@link Source} and {@link Sink} classes and implements common logic. It's required to have separate classes
+ * for Source and Sink in the case of different schema readers.
+ */
+public abstract class SqlServerDBRecord extends DBRecord {
+
+  public SqlServerDBRecord(StructuredRecord record, int[] columnTypes) {
+    super(record, columnTypes);
+  }
+
+  public SqlServerDBRecord() {}
+
+  @Override
+  protected void handleField(ResultSet resultSet, StructuredRecord.Builder recordBuilder, Schema.Field field,
+                             int columnIndex, int sqlType, int sqlPrecision, int sqlScale) throws SQLException {
+    switch (sqlType) {
+      case Types.TIME:
+        // Handle reading SQL Server 'TIME' data type to avoid accuracy loss.
+        // 'TIME' data type has the accuracy of 100 nanoseconds(1 millisecond in Informatica)
+        // but reading via 'getTime' and 'getObject' will round value to second.
+        recordBuilder.setTime(field.getName(), resultSet.getTimestamp(columnIndex).toLocalDateTime().toLocalTime());
+        break;
+      case SqlServerSchemaReader.DATETIME_OFFSET_TYPE:
+        recordBuilder.set(field.getName(), resultSet.getString(columnIndex));
+        break;
+      default:
+        setField(resultSet, recordBuilder, field, columnIndex, sqlType, sqlPrecision, sqlScale);
+    }
+  }
+
+  /**
+   * SQL Server Source implementation of {@link org.apache.hadoop.mapreduce.lib.db.DBWritable} and
+   * {@link org.apache.hadoop.io.Writable}.
+   */
+  public static class Source extends SqlServerDBRecord {
```

Comment:
This doesn't look like it behaves any different than SqlServerDBRecord, can just make the parent non-abstract.

Reply:
Done.